applications/serial_lte_modem/src/slm_cmux.c

```diff
@@ -50,40 +54,87 @@ static struct {
 	} dlcis[CHANNEL_COUNT];
 	/* Index of the DLCI used for AT communication; defaults to 0. */
 	unsigned int at_channel;
+
+	/* Incoming data for DLCI's. */
+	atomic_t dlci_channel_rx;
+	struct k_work rx_work;
+
+	/* Outgoing data for AT DLCI. */
+	struct k_sem tx_sem;
+	struct ring_buf notif_rb;
+	uint8_t notif_buffer[CONFIG_SLM_CMUX_NOTIFICATION_TX_BUFFER_SIZE];
+	struct k_mutex notif_rb_mutex;
+	struct k_work tx_work;
+
 } cmux;
 
+static void rx_work_fn(struct k_work *work)
+{
+	static uint8_t recv_buf[RECV_BUF_LEN];
+	int ret;
+	bool is_at;
+
+	for (int i = 0; i < ARRAY_SIZE(cmux.dlcis); ++i) {
+		if (atomic_test_and_clear_bit(&cmux.dlci_channel_rx, i)) {
+			/* Incoming data for DLCI. */
+			is_at = (i == cmux.at_channel);
+
+			ret = modem_pipe_receive(cmux.dlcis[i].pipe, recv_buf, sizeof(recv_buf));
+			if (ret < 0) {
+				LOG_ERR("%sDLCI %u failed modem_pipe_receive. (%d)",
+					is_at ? "AT " : "", INDEX_TO_DLCI(i), ret);
+				continue;
+			}
+
+			if (!is_at) {
+				LOG_INF("DLCI %u discarding %u bytes of data.", INDEX_TO_DLCI(i),
+					ret);
+				continue;
+			}
+
+			LOG_DBG("AT DLCI %u received %u bytes of data.", INDEX_TO_DLCI(i), ret);
+			slm_at_receive(recv_buf, ret);
+		}
+	}
+}
+
 static void dlci_pipe_event_handler(struct modem_pipe *pipe,
 				    enum modem_pipe_event event, void *user_data)
 {
 	const struct cmux_dlci *const dlci = user_data;
+	bool is_at = (ARRAY_INDEX(cmux.dlcis, dlci) == cmux.at_channel);
 
 	switch (event) {
-	case MODEM_PIPE_EVENT_OPENED:
-	case MODEM_PIPE_EVENT_CLOSED:
 		/* The events of the DLCIs other than that of the AT channel
 		 * are received here when they haven't been attached to
 		 * by their respective implementations.
 		 */
-		LOG_INF("DLCI %u %s.", dlci->address,
-			(event == MODEM_PIPE_EVENT_OPENED) ? "opened" : "closed");
+	case MODEM_PIPE_EVENT_OPENED:
+		LOG_INF("%sDLCI %u opened.", is_at ? "AT " : "", dlci->address);
```

Comment:
This way of logging may give an impression like there would be many DLCIs for AT for an inexperienced reader of the logging.
Another way of emphasizing the AT channel could be as follows:
```suggestion
		LOG_INF("DLCI %u %s opened.", dlci->address, is_at ? "(AT) " : "");
```

This is similarly in few other places.